// @microsoft/api-extractor/lib/aedoc/ReleaseTag.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/**
  * A "release tag" is a custom TSDoc tag that is applied to an API to communicate the level of support
  * provided for third-party developers.
  *
  * @remarks
  *
  * The four release tags are: `@internal`, `@alpha`, `@beta`, and `@public`. They are applied to API items such
  * as classes, member functions, enums, etc.  The release tag applies recursively to members of a container
  * (e.g. class or interface). For example, if a class is marked as `@beta`, then all of its members automatically
  * have this status; you DON'T need add the `@beta` tag to each member function. However, you could add
  * `@internal` to a member function to give it a different release status.
  *
  * @public
  */
var ReleaseTag;
(function (ReleaseTag) {
    /**
     * No release tag was specified in the AEDoc summary.
     */
    ReleaseTag[ReleaseTag["None"] = 0] = "None";
    /**
     * Indicates that an API item is meant only for usage by other NPM packages from the same
     * maintainer. Third parties should never use "internal" APIs. (To emphasize this, their
     * names are prefixed by underscores.)
     */
    ReleaseTag[ReleaseTag["Internal"] = 1] = "Internal";
    /**
     * Indicates that an API item is eventually intended to be public, but currently is in an
     * early stage of development. Third parties should not use "alpha" APIs.
     */
    ReleaseTag[ReleaseTag["Alpha"] = 2] = "Alpha";
    /**
     * Indicates that an API item has been released in an experimental state. Third parties are
     * encouraged to try it and provide feedback. However, a "beta" API should NOT be used
     * in production.
     */
    ReleaseTag[ReleaseTag["Beta"] = 3] = "Beta";
    /**
     * Indicates that an API item has been officially released. It is part of the supported
     * contract (e.g. SemVer) for a package.
     */
    ReleaseTag[ReleaseTag["Public"] = 4] = "Public";
})(ReleaseTag = exports.ReleaseTag || (exports.ReleaseTag = {}));
