// @microsoft/api-extractor/lib/collector/DeclarationMetadata.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const ReleaseTag_1 = require("../aedoc/ReleaseTag");
class DeclarationMetadata {
    constructor() {
        this.tsdocComment = undefined;
        this.tsdocParserContext = undefined;
        /**
         * This is the release tag that was explicitly specified in the original doc comment, if any.
         * Compare with SymbolMetadata.releaseTag, which is the effective release tag, possibly inherited from
         * a parent.
         */
        this.declaredReleaseTag = ReleaseTag_1.ReleaseTag.None;
        // NOTE: In the future, the Collector may infer or error-correct some of these states.
        // Generators should rely on these instead of tsdocComment.modifierTagSet.
        this.isEventProperty = false;
        this.isOverride = false;
        this.isSealed = false;
        this.isVirtual = false;
        this.needsDocumentation = true;
    }
}
exports.DeclarationMetadata = DeclarationMetadata;
