// @microsoft/api-extractor/lib/api/mixins/Excerpt.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/** @public */
class ExcerptToken {
    constructor(kind, text) {
        this.kind = kind;
        this.text = text;
    }
}
exports.ExcerptToken = ExcerptToken;
/**
 * This class is used by {@link (ApiDeclaredItem:interface)} to represent a source code excerpt containing
 * a TypeScript declaration.
 *
 * @remarks
 *
 * The main excerpt is parsed into an array of tokens, and the main excerpt's token range will span all of these
 * tokens.  The declaration may also have have "captured" excerpts, which are other subranges of tokens.
 * For example, if the main excerpt is a function declaration, it will also have a captured excerpt corresponding
 * to the return type of the function.
 *
 * An excerpt may be empty (i.e. a token range containing zero tokens).  For example, if a function's return value
 * is not explicitly declared, then the returnTypeExcerpt will be empty.  By contrast, a class constructor cannot
 * have a return value, so ApiConstructor has no returnTypeExcerpt property at all.
 *
 * @public
 */
class Excerpt {
    constructor(tokens, tokenRange) {
        this.tokens = tokens;
        this.tokenRange = tokenRange;
        if (this.tokenRange.startIndex < 0 || this.tokenRange.endIndex > this.tokens.length
            || this.tokenRange.startIndex > this.tokenRange.endIndex) {
            throw new Error('Invalid token range');
        }
    }
    get text() {
        if (this._text === undefined) {
            this._text = this.tokens.slice(this.tokenRange.startIndex, this.tokenRange.endIndex)
                .map(x => x.text).join('');
        }
        return this._text;
    }
}
exports.Excerpt = Excerpt;
