// @microsoft/api-extractor/lib/analyzer/AstModule.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
class AstModuleExportInfo {
    constructor() {
        this.exportedLocalSymbols = new Map();
        this.starExportedExternalModules = new Set();
    }
}
exports.AstModuleExportInfo = AstModuleExportInfo;
/**
 * An internal data structure that represents a source file that is analyzed by AstSymbolTable.
 */
class AstModule {
    constructor(sourceFile, moduleSymbol, externalModulePath) {
        this.sourceFile = sourceFile;
        this.moduleSymbol = moduleSymbol;
        this.externalModulePath = externalModulePath;
        this.starExportedModules = new Set();
        this.cachedExportedSymbols = new Map();
        this.astModuleExportInfo = undefined;
    }
    get isExternal() {
        return this.externalModulePath !== undefined;
    }
}
exports.AstModule = AstModule;
